users/views.py

```diff
@@ -211,32 +214,111 @@ def get_active_social_users(self, user):
             return []
         return UserSocialAuth.objects.filter(user=user)
 
+    def get_oidc_backends_end_session_url(self, request, social_users):
+        """Return end session url of the first OIDC backend the user has a
+        social auth entry on and hasn't been redirected yet.
+
+        The already redirected backends are tracked in a session variable because
+        in theory there could be multiple OIDC social auths for a user.
+        """
+        for social_user in social_users:
+            try:
+                backend_class = get_backend_class(social_user.provider)
+            except MissingBackend:
+                logger.warning(
+                    'Missing backend "{}" in user social auths for user {}'.format(
+                        social_user.provider,
+                        request.user.pk
+                    )
+                )
+                continue
+
+            if not issubclass(backend_class, OpenIdConnectAuth):
+                continue
+
+            backend = backend_class()
+
+            # Only redirect if it's enabled in the backend
+            if not backend.setting('REDIRECT_LOGOUT_TO_END_SESSION', default=False):
+                continue
+
+            end_session_endpoint = backend.oidc_config().get('end_session_endpoint')
+            if not end_session_endpoint:
+                continue
+
+            if 'oidc_backend_end_session_redirected' not in request.session:
+                request.session['oidc_backend_end_session_redirected'] = defaultdict(
+                    bool
+                )
+
+            if request.session['oidc_backend_end_session_redirected'].get(backend.name):
+                continue
+
+            request.session['oidc_backend_end_session_redirected'][backend.name] = True
+
+            end_session_url = '{}?{}'.format(end_session_endpoint, urlencode({
+                'post_logout_redirect_uri': request.build_absolute_uri(request.path)
+            }))
+
+            return end_session_url
+
     def dispatch(self, request, *args, **kwargs):
-        # Check if the authenticated user has active Suomi.fi login
         user = request.user
-        social_user = self._active_suomi_fi_social_user(user)
+
+        social_users = self.get_active_social_users(user)
+        if len(social_users) > 1:
+            logger.warning(
+                'Multiple active social core backends for user {}'.format(
+                    user.pk
+                )
+            )
```

Comment:
Does this need to be logged (I don't know the logging practices in this project, so it can well be) and is warning the correct level? At least the multiple social users case is now handled nicely, so what's there to report? Or perhaps put level to info or something like that?

Reply:
Don't know. It was there already. I just moved it.

Reply:
Info might be better. The multiple social user case is now handled so that there will be a new user for each backend. Still there might be old users that have ended up with multiple social backends. But still, that will have to be handled once we start migrating, and I guess this is the wrong place to complain about it, it's more like a general database data quality issue. Removing is ok by me.